src/pyuvdata/uvbeam/analytic_beam.py

```diff
@@ -0,0 +1,893 @@
+            model_version="1.0",
+            freq_array=freq_array,
+            feed_array=feed_array,
+            polarization_array=polarization_array,
+            x_orientation=self.x_orientation,
+            pixel_coordinate_system=pixel_coordinate_system,
+            axis1_array=axis1_array,
+            axis2_array=axis2_array,
+            nside=nside,
+            healpix_pixel_array=healpix_pixel_array,
+            ordering=ordering,
+            history=f"created from a pyuvdata analytic beam: {self.__repr__()}",
+        )
+
+        if uvb.pixel_coordinate_system == "healpix":
+            try:
+                from astropy_healpix import HEALPix
+            except ImportError as e:
+                raise ImportError(
+                    "astropy_healpix is not installed but is "
+                    "required for healpix functionality. "
+                    "Install 'astropy-healpix' using conda or pip."
+                ) from e
+            hp_obj = HEALPix(nside=uvb.nside, order=uvb.ordering)
+            hpx_lon, hpx_lat = hp_obj.healpix_to_lonlat(uvb.pixel_array)
+            za_array, az_array = utils.coordinates.hpx_latlon_to_zenithangle_azimuth(
+                hpx_lat.radian, hpx_lon.radian
+            )
+
+        else:
+            az_array, za_array = np.meshgrid(uvb.axis1_array, uvb.axis2_array)
+            az_array = az_array.flatten()
+            za_array = za_array.flatten()
+
+        if beam_type == "efield":
+            eval_function = "efield_eval"
+        else:
+            eval_function = "power_eval"
+
+        data_array = getattr(self, eval_function)(
+            az_array=az_array, za_array=za_array, freq_array=freq_array
+        )
+
+        if uvb.pixel_coordinate_system == "az_za":
+            data_array = data_array.reshape(uvb.data_array.shape)
+
+        uvb.data_array = data_array
+
+        uvb.check()
+        return uvb
+
+
+def analytic_beam_constructor(loader, node):
+    """
+    Define a yaml constructor for analytic beams.
+
+    The yaml must specify a "class" field with an importable class and any
+    required inputs to that class's constructor.
+
+    Parameters
+    ----------
+    loader: yaml.Loader
+        An instance of a yaml Loader object.
+    node: yaml.Node
+        A yaml node object.
+
+    Returns
+    -------
+    beam
+        An instance of an AnalyticBeam subclass.
+
+    """
+    values = loader.construct_mapping(node)
+    if "class" not in values:
+        raise ValueError("yaml entries for AnalyticBeam must specify a class")
+    class_parts = (values.pop("class")).split(".")
+    class_name = class_parts[-1]
+    if len(class_parts) == 1:
+        # no module specified, assume pyuvdata
+        module = importlib.import_module("pyuvdata")
+    else:
+        module = (".").join(class_parts[:-1])
+        module = importlib.import_module(module)
+    beam_class = getattr(module, class_name)
+
+    beam = beam_class(**values)
```

Comment:
I think we can offer slightly more ease-of-use by also implementing the plugin structure I mentioned previously. So, implement the following magic method on this class:

```python
class AnalyticBeam:
    __types__: dict = {}

    def __init_subclass__(cls):
        # consider using cls.name.lower()
        cls.__types__[cls.name] = cls
```

Then in this method, we do something similar to what you have:

```python
if len(class_parts) > 1:
    module = (".").join(class_parts[:-1])
    importlib.import_module(module)
 
# any  subclass that has been imported will work (obviously including the pyuvdata ones...)
beam_class = AnalyticBeam.__types__[class_name]
beam = beam_class(**values)
```

Reply:
This also then means we can do more helpful validation:

```python
if class_name not in `AnalyticBeam.__types__`:
    raise NameError(f"{class_name} is not a known AnalyticBeam. Available options are: {list(AnalyticBeam.__types__.keys())}. If it is a custom beam, either ensure the module is imported, or specify the beam with dot-pathed modules included (i.e. `my_module.MyAnalyticBeam`)")
```

Reply:
I don't think that last point will work. It seems like the `__init_subclass__` is only run when AnalyticBeam is imported, so if the new module is imported afterwards (e.g. in the function above where this is called), it won't exist in the `AnalyticBeam.__types__`. I even tried adding `AnalyticBeam.__init_subclass__()` after the new import and it doesn't update the `AnalyticBeam.__types__`.

Reply:
I can put the whole earlier block in a try/except and use this message as the final error in the except.

Reply:
Hmmm, I'm not understanding what won't work here. I've definitely used this pattern before, but maybe I'm misunderstanding what you're saying. Any subclass that is imported at any time will add itself to `__types__` upon definition.

Reply:
Maybe the problem is how I'm trying to test it. I just pushed a commit that adds this infrastructure and adds a test that works without the final check but breaks with it.

I order to access the part of the constructor that works imports a class if it isn't already imported, I set `AnalyticBeam.__types__={}`. That ensures that I enter that part of the code but then it errors on the final check.

Reply:
I think this is sorted out now, the problem was in the test, I came up with a better approach for that.